### PR TITLE
Mem UI

### PR DESCRIPTION
**Pull request by ThePythonGuy3**
Adds a simple UI to check the memory cell contents. [Image](https://discord.com/channels/391020510269669376/663853133587546138/759503028985790494)

### CODE REVIEW

**Comment by ThePythonGuy3** [comment]
~~Yes i love messing with ui stuff~~

File: core/src/mindustry/world/blocks/logic/MemoryBlock.java [hunk 1]
```diff
@@ -1,17 +1,24 @@
 package mindustry.world.blocks.logic;
 
+import arc.scene.ui.layout.Table;
+import arc.util.Align;
 import arc.util.io.*;
 import mindustry.gen.*;
+import mindustry.ui.Styles;
+import mindustry.ui.dialogs.BaseDialog;
```
Comment (summetdev) [inline core/src/mindustry/world/blocks/logic/MemoryBlock.java:8]: Use `.*` imports.

File: core/src/mindustry/world/blocks/logic/MemoryBlock.java [hunk 2]
```diff
@@ -1,17 +1,24 @@
...
 import mindustry.world.*;
 import mindustry.world.meta.*;
 
+import java.awt.*;
```
Comment (summetdev) [inline core/src/mindustry/world/blocks/logic/MemoryBlock.java:12]: Don't use [incompatible Java features](https://github.com/Anuken/Mindustry/blob/master/CONTRIBUTING.md#do-not-use-incompatible-java-features-javautilfunction-javaawt).

File: core/src/mindustry/world/blocks/logic/MemoryBlock.java [hunk 3]
```diff
@@ -21,9 +28,21 @@ public void setStats(){
         stats.add(BlockStat.memoryCapacity, memoryCapacity, StatUnit.none);
     }
 
-    public class MemoryBuild extends Building{
+    public class MemoryBuild extends Building implements mindustry.world.blocks.logic.MemoryBuild {
```
Comment (summetdev) [inline core/src/mindustry/world/blocks/logic/MemoryBlock.java:31]: Why not import it?

Comment (ThePythonGuy3) [reply core/src/mindustry/world/blocks/logic/MemoryBlock.java:31]: what i dont get it

File: core/src/mindustry/world/blocks/logic/MemoryBlock.java (hunk 1 above)
Comment (ThePythonGuy3) [reply core/src/mindustry/world/blocks/logic/MemoryBlock.java:8]: i absolutely have no idea on why Intellij goes so mad

**Comment by ThePythonGuy3** [comment]
Fixed

File: core/src/mindustry/world/blocks/logic/MemoryBlock.java [hunk 4]
```diff
@@ -21,9 +25,21 @@ public void setStats(){
         stats.add(BlockStat.memoryCapacity, memoryCapacity, StatUnit.none);
     }
 
-    public class MemoryBuild extends Building{
+    public class MemoryBuild extends Building {
```
Comment (l-Luna) [inline core/src/mindustry/world/blocks/logic/MemoryBlock.java:28]: why add a space

**Review by l-Luna — APPROVED** [review]

File: core/src/mindustry/world/blocks/logic/MemoryBlock.java (hunk 4 above)
Comment (summetdev) [reply core/src/mindustry/world/blocks/logic/MemoryBlock.java:28]: ~~Anuke doesn't like spaces.~~

**Comment by DeltaNedas** [comment]
deltanedas/ldb stolen

File: core/src/mindustry/world/blocks/logic/MemoryBlock.java (hunk 4 above)
Comment (ThePythonGuy3) [reply core/src/mindustry/world/blocks/logic/MemoryBlock.java:28]: No idea

**Comment by Anuken** [comment]
There are several problems with this:
- It's not dynamic, which means any changes to memory while configuring would not be reflected
  - Making it dynamic would mean spewing garbage everywhere and lagging even more
- It's not very easy to use
- It's going to lag, because 500+ labels in a row is not great for mobile performance, especially considering they're set up every time you click on a memory block
- Is it really necessary? If you really want to debug, why not get and print?

**Comment by sk7725** [comment]
Add a "refresh" button at the top.
This solves:
> It's going to lag, because 500+ labels in a row is not great for mobile performance, especially considering they're set up every time you click on a memory block
And also implies than you have to refresh it yourself in a simple way:
> It's not dynamic, which means any changes to memory while configuring would not be reflected
Also maybe a dedicated FloatingDialog will suit better than just a table.

**Comment by ThePythonGuy3** [comment]
hmmm i will take a look at those points

**Comment by ThePythonGuy3** [comment]
@Anuken now i fixed the things you and others pointed out. [Final result](https://discord.com/channels/391020510269669376/663853133587546138/761301438981996564)

**Comment by Anuken** [comment]
> Add a "refresh" button at the top.
This solves:

Well, no, it doesn't, because you still have to set the labels up and check their culling every frame.

The only valid implementation for this that I would accept is a custom element that:
- doesn't use any elements, only one display
- directly draws using the font
- only displays the numbers in range of the screen by calculating the culling range (nothing outside the range is updated, created or drawn)

**Pull request closed without merging**